kubetest2/kubetest2-gke/deployer/deployer.go

```diff
@@ -0,0 +1,516 @@
+		"--project="+d.project,
+		location)...))
+	if err != nil {
+		return fmt.Errorf("instance group URL fetch failed: %s", execError(err))
+	}
+	igURLs := strings.Split(strings.TrimSpace(string(igs)), ";")
+	if len(igURLs) == 0 {
+		return fmt.Errorf("no instance group URLs returned by gcloud, output %q", string(igs))
+	}
+	sort.Strings(igURLs)
+	for _, igURL := range igURLs {
+		m := poolRe.FindStringSubmatch(igURL)
+		if len(m) == 0 {
+			return fmt.Errorf("instanceGroupUrl %q did not match regex %v", igURL, poolRe)
+		}
+		d.instanceGroups = append(d.instanceGroups, &ig{path: m[0], zone: m[1], name: m[2], uniq: m[3]})
+	}
+	return nil
+}
+
+func (d *deployer) getClusterFirewall() (string, error) {
+	if err := d.getInstanceGroups(); err != nil {
+		return "", err
+	}
+	// We want to ensure that there's an e2e-ports-* firewall rule
+	// that maps to the cluster nodes, but the target tag for the
+	// nodes can be slow to get. Use the hash from the lexically first
+	// node pool instead.
+	return "e2e-ports-" + d.instanceGroups[0].uniq, nil
+}
+
+// This function ensures that all firewall-rules are deleted from specific network.
+// We also want to keep in logs that there were some resources leaking.
+func (d *deployer) cleanupNetworkFirewalls() (int, error) {
+	fws, err := exec.Output(exec.Command("gcloud", "compute", "firewall-rules", "list",
+		"--format=value(name)",
+		"--project="+d.project,
+		"--filter=network:"+d.network))
+	if err != nil {
+		return 0, fmt.Errorf("firewall rules list failed: %s", execError(err))
+	}
+	if len(fws) > 0 {
+		fwList := strings.Split(strings.TrimSpace(string(fws)), "\n")
+		log.Printf("Network %s has %v undeleted firewall rules %v", d.network, len(fwList), fwList)
+		commandArgs := []string{"compute", "firewall-rules", "delete", "-q"}
+		commandArgs = append(commandArgs, fwList...)
+		commandArgs = append(commandArgs, "--project="+d.project)
+		errFirewall := runWithOutput(exec.Command("gcloud", commandArgs...))
+		if errFirewall != nil {
+			return 0, fmt.Errorf("error deleting firewall: %v", errFirewall)
+		}
+		return len(fwList), nil
+	}
+	return 0, nil
+}
+
+func (d *deployer) Down() error {
+	if err := d.verifyFlags(); err != nil {
+		return err
+	}
+	if err := d.prepareGcpIfNeeded(); err != nil {
+		return err
+	}
+	firewall, err := d.getClusterFirewall()
+	if err != nil {
+		// This is expected if the cluster doesn't exist.
+		return nil
+	}
+	d.instanceGroups = nil
+
+	loc, err := d.location()
+	if err != nil {
+		return err
+	}
+
+	// We best-effort try all of these and report errors as appropriate.
+	errCluster := runWithOutput(exec.Command(
+		"gcloud", d.containerArgs("clusters", "delete", "-q", d.cluster,
+			"--project="+d.project,
+			loc)...))
+
+	// don't delete default network
+	if d.network == "default" {
+		if errCluster != nil {
+			log.Printf("Error deleting cluster using default network, allow the error for now %s", errCluster)
+		}
+		return nil
+	}
+
+	var errFirewall error
```

Comment:
we should consider breaking some of these out into higher level subroutines in the future like "delete firewalls"